### DISCUSSION TITLE
Paranoid mode: How to redirect from "unsafe" asdf: community plugin to safe ubi: alternative

### DISCUSSION THREAD

**Discussion by powerman** [jdx/mise#2878] [Q&A]
Let's say I'm trying to use paranoid mode and need to work on some (not mine own) project which already uses tool "mage" (which is a shorthand in registry for "asdf:mathew-fleisch/asdf-mage").

Is it possible in my local.toml or my global config to somehow redirect tool name "mage" to safer "ubi:magefile/mage"? I mean I can't just replace tool name in that project's `.mise.toml` because it'll be complicated to make sure this change won't be occasionally pushed to other developers.

**Comment by jdx** [comment]
If it works in ubi it should be the default for everyone, not just when paranoid is enabled

**Reply by RafalSkolasinski** [reply]
Is there some simple command that would allow to drop asdf plugins in favour of ubi whenever possible?

**Reply by jdx** [reply]
that is the default behavior

**Reply by RafalSkolasinski** [reply]
So if I had in example `kind` installed through asdf-vm plugin, it will automatically be replaced by binaries coming from `ubi`?

**Reply by jdx** [reply]
if it's in registry.toml, yes, otherwise mise would have no idea how to map `asdf:kind` -> `ubi:kubernetes-sigs/kind`

**Reply by RafalSkolasinski** [reply]
I guess that's kind of what I was asking. I have setup migrated some time ago from asdf with quite a few plugins installed. 
Now I just became aware about ubi / aqua backends and they seem much more secure to community written asdf plugins. 

Maybe I should RTFM on changes in latest versions ;) - but first search lead me here... 

I just dropped all my plugins and removed tools
```
mise plugins remove --all
mise uninstall --all
```

and then after doing 
```
mise install
``` 
it only asked me to add few plugins for tools that indeed has no ubi/aqua options. 


I guess that's a `poor man` solution to what I wanted to did :). Without nee to edit any configs etc.

**Comment by powerman** [comment]
It's not the only one in current registry which can be replaced with ubi - here is the full list:
```
evans = "ubi:ktr0731/evans"
gitleaks = "ubi:gitleaks/gitleaks"
go-swagger = "ubi:go-swagger/go-swagger"
gotestsum = "ubi:gotestyourself/gotestsum"
grpcurl = "ubi:fullstorydev/grpcurl"
jqp = "ubi:noahgorstein/jqp"
mage = "ubi:magefile/mage"
minikube = "ubi:kubernetes/minikube"
stylua = "ubi:JohnnyMorganz/stylua"
task = "ubi:go-task/task"
yamlfmt = "ubi:google/yamlfmt"
```

**Reply by powerman** [reply]
One of above needs extra `exe` option:
```
"ubi:go-swagger/go-swagger" = {  exe = "swagger" }
```

**Reply by BurnerWah** [reply]
in my testing gitleaks & stylua appeared to run into issues on less common architectures, so not all of these are totally viable to add UBI installers for as the default (at least not yet).
I can test some of the other things listed here next week

**Reply by powerman** [reply]
> in my testing gitleaks & stylua appeared to run into issues on less common architectures, so not all of these are totally viable to add UBI installers for as the default (at least not yet).

In theory - you're right, asdf plugins may support extra edge cases.
In practice - asdf plugins for both these tools are installing binary gh release, just like ubi.

**Reply by BurnerWah** [reply]
> > in my testing gitleaks & stylua appeared to run into issues on less common architectures, so not all of these are totally viable to add UBI installers for as the default (at least not yet).
> 
> In theory - you're right, asdf plugins may support extra edge cases. In practice - asdf plugins for both these tools are installing binary gh release, just like ubi.

Yeah, the ASDF plugins are doing what UBI is doing with extra steps. The problem is that UBI is prone to not failing when it should, while ASDF plugins, being specialized for specific tools, can actually fail on unsupported platforms.
Gitleaks and stylua specifically ran into issues with installing aarch64 binaries on architectures like risc-v and PowerPC (I need to report this but I need proper debug logs first), and I've had some other edge cases that have kept me from including semi-functional UBI installerd for tools.
In these cases, it's preferable for the backend to fail outright instead of it downloading impossible to execute binaries. 

As UBI's edge cases get sorted out, it'll become more and more practical as the default method to install things, but until then, I'm rather hesitant on making it the default for things when I know it won't work correctly. I'm fine with using the UBI backend for stylua in my own projects, but I don't think that "works on my system" is good enough criteria for the default on all systems.

**Reply by powerman** [reply]
Just curious: is https://github.com/zyedidia/eget works in these edge cases where ubi fails?

**Reply by BurnerWah** [reply]
> Just curious: is https://github.com/zyedidia/eget works in these edge cases where ubi fails?

I haven't checked yet, since there isn't an eget backend available for mise.
Even if it does, UBI is built into mise itself, while eget would probably be harder to embed into mise.
I could try and see if it works when I get back to testing stuff next week though.

**Reply by jdx** [reply]
does mise itself work on PowerPC?

**Reply by BurnerWah** [reply]
> does mise itself work on PowerPC?

I actually haven't done much testing of mise itself on other architectures, but I could probably test to see if mise can build on PowerPC and risc-v later today
I don't think I'll be able to get a VM for testing the whole program on PowerPC running until Monday, but I do have a risc-v VM on hand for testing already

**Reply by BurnerWah** [reply]
> does mise itself work on PowerPC?

After some testing, for the most part it does work on PowerPC (specifically ppc64le, running on fedora 41 since the ubuntu rustc is too old).
The cargo backend is basically non-functional, probably due to cargo-binstall and sccache not having any prebuilt binaries available and me being unable to compile release builds for anything at all.
Otherwise, the issues I've run into seem like they'd be pretty easy to fix.
The python, node, go, and zig plugins all just need some code to change the architecture they try to download from `powerpc64` to something like `ppc64le` or `powerpc64le`
And the java plugin has no metadata to work off of, so it fails to function outright, but I know temurin has ppc64 and ppc64le java available, so it could be made to work.
Deno & Bun have no prebuilt versions available.

As for tests, I got all but three unit tests to pass.

So yeah for the most part, it's functional.

**Reply by jdx** [reply]
> The cargo backend is basically non-functional, probably due to cargo-binstall and sccache not having any prebuilt binaries available and me being unable to compile release builds for anything at all.

That should only be a problem if actually trying to use cargo-binstall and sccache. If you don't put them into your config mise won't try to install them.

**Comment by powerman** [comment]
Anyway, looks like my original question should be reworded in this way: if I'll see some `asdf:` plugin in 3rd-party project's `.mise.toml` in paranoid mode, is it possible to somehow replace it with my own (reviewed and not updated without my control - i.e. my own fork of that asdf plugin) asdf plugin, so that original `.mise.toml` will continue to work without modifications but my local system will be protected from executing non-reviewed 3rd-party code?

**Reply by powerman** [reply]
Maybe will be solved by https://github.com/jdx/mise/issues/2885

**Comment by JamBalaya56562** [comment]
Your reworded question — replacing a tool in a third-party project's `mise.toml` with something you control, **without editing that file** — has an answer now.

**`[tool_alias]` in your global config redirects the tool name for the whole machine.** Measured on **v2026.8.3** with `~/.config/mise/config.toml`:

```toml
[tool_alias]
mage = "github:magefile/mage"
```

and an untouched project config containing `mage = "latest"`:

```console
$ mise tool mage
Backend:            github:magefile/mage
Active Version:     1.17.2
```

Point it at your own fork the same way — `asdf:you/your-reviewed-asdf-mage` — and the project's file keeps working unmodified while your machine never runs the upstream plugin.

**Two traps worth knowing.**

`[plugins]` is *not* this, despite looking like it. `[plugins] mage = "github:magefile/mage"` is read as an **asdf plugin source**, so you get `Backend: asdf:github:magefile/mage` and then `[mage] plugin not installed`. Plain `[alias]` still works but warns — `[alias] is deprecated, use [tool_alias] instead`.

**The `ubi:` targets in my list above are stale.** `ubi` is deprecated in favour of `github:` — *"The ubi backend is deprecated. Use the github backend instead (e.g., github:owner/repo)"* — warning since **2026.4.0** and removed in **2027.1.0**. So `ubi:magefile/mage` should read `github:magefile/mage` today.

Also worth checking before doing any of this: **`mage` already resolves to `aqua:magefile/mage`**, not the `asdf:` this thread assumes. Several entries on that list may have moved the same way, which would make a redirect unnecessary for them — `mise tool <name>` shows the backend a shorthand resolves to.
